airflow/contrib/operators/gcp_compute_operator.py

```diff
@@ -0,0 +1,186 @@
+        if not self.zone:
+            raise AirflowException("The required parameter 'zone' is missing")
+        if not self.resource_id:
+            raise AirflowException("The required parameter 'resource_id' is missing")
+
+    def execute(self, context):
+        return self._hook.stop_instance(self.project_id, self.zone, self.resource_id)
+
+
+SET_MACHINE_TYPE_VALIDATION_SPECIFICATION = [
+    dict(name="machineType", regexp="^.+$"),
+]
+
+
+class GceSetMachineTypeOperator(BaseOperator):
+    """
+    Changes the machine type for a stopped instance to the machine type specified in
+    the request.
+
+    :param project_id: Google Cloud Platform project where the Compute Engine
+        instance exists.
+    :type project_id: str
+    :param zone: Google Cloud Platform zone where the instance exists.
+    :type zone: str
+    :param resource_id: Name of the Compute Engine instance resource.
+    :type resource_id: str
+    :param body: Body passed to the setMachineType method call.
+    :type body: dict
+    :param gcp_conn_id: The connection ID used to connect to Google Cloud Platform.
+    :type gcp_conn_id: str
+    :param api_version: API version used (e.g. v1).
+    :type api_version: str
+    """
+    @apply_defaults
+    def __init__(self,
+                 project_id,
+                 zone,
+                 resource_id,
+                 body,
+                 gcp_conn_id='google_cloud_default',
+                 api_version='v1',
+                 validate_body=True,
+                 *args, **kwargs):
+        self.project_id = project_id
+        self.zone = zone
+        self.full_location = 'projects/{}/zones/{}'.format(self.project_id,
+                                                           self.zone)
+        self.resource_id = resource_id
+        self.gcp_conn_id = gcp_conn_id
+        self.api_version = api_version
+        self.body = body
+        self._field_validator = None
+        if validate_body:
+            self._field_validator = BodyFieldValidator(
+                SET_MACHINE_TYPE_VALIDATION_SPECIFICATION, api_version=api_version)
+        self._validate_inputs()
+        self._hook = GceHook(gcp_conn_id=self.gcp_conn_id, api_version=self.api_version)
+        super(GceSetMachineTypeOperator, self).__init__(*args, **kwargs)
+
+    def _validate_all_body_fields(self):
+        if self._field_validator:
+            self._field_validator.validate(self.body)
+
+    def _validate_inputs(self):
+        if not self.project_id:
+            raise AirflowException("The required parameter 'project_id' is missing")
+        if not self.zone:
+            raise AirflowException("The required parameter 'zone' is missing")
+        if not self.resource_id:
+            raise AirflowException("The required parameter 'resource_id' is missing")
+
+    def execute(self, context):
+        self._validate_all_body_fields()
+        return self._hook.set_machine_type(self.project_id, self.zone,
```

Comment:
Please confirm this is idempotent

Reply:
Confirmed as described in https://github.com/PolideaInternal/incubator-airflow/pull/4#discussion_r222334621